### ISSUE TITLE
output reports on the CLI

### ISSUE THREAD

**Issue by kronn** [gtimelog/gtimelog#43]
I love the simplicity to enter things via the GUI. However, I do not mail my reports to anyone anymore. I did that for a while and it was nice and easy with gtimelog.

I am looking for a way to get the report for a week output on the CLI, without the whole mail-overhead. 

Is something like this planned?

I am willing to contribute, though I have not much python experience. If this is not planned, I would be happy about some pointer where and how to start.

**Comment by kronn** [comment]
I can see that solving this would also solve https://bugs.launchpad.net/gtimelog/+bug/248612.

**Comment by mgedmin** [comment]
Curiously cmdline-only reports [were there in the beginning](https://github.com/gtimelog/gtimelog/tree/master/scripts), before any GUI stuff.  I suppose those scripts still work, since the format of `timelog.txt` remains unchanged.

I also hate that I have to spawn a mail client (and then abort the mail sending) if all I want to do is look at a report.  History browsing fixes this for daily reports, somewhat.  I would like to make it possible to see weekly/monthly summaries in the GUI too.  I don't know if that would help you, or if you really want a command-line version.

Writing a CLI report generator using the internal gtimelog.timelog module should be trivial -- but isn't:
```python
#!/usr/bin/env python
"""Print the weekly report for this week to stdout."""
import datetime
import os
import sys
from gtimelog.timelog import TimeLog, Reports

filename = os.path.expanduser('~/.local/share/gtimelog/timelog.txt') # XXX: could be ~/.gtimelog/timelog.txt
virtual_midnight = datetime.time(2, 0) # XXX hardcoded, might not match ~/.config/gtimelog/gtimelogrc
email = ''  # will be used for printing the 'To: <email>' line to stdout
name = 'myself'  # will be used for printing the 'Subject: weekly report for <name> (...)' line

if __name__ == '__main__':
    log = TimeLog(filename, virtual_midnight)
    report = Reports(log.window_for_week(datetime.date.today()))
    report.weekly_report(sys.stdout, email, name)
```

The APIs of gtimelog.timelog are internal and might change (I hope so!), but I've little time/inspiration to work on gtimelog, so the change isn't likely to be fast and painful :)

**Comment by mgedmin** [comment]
I wouldn't object to a pull request for letting gtimelog generate the reports using the CLI.  A possible UI:
```
$ gtimelog --report [--daily|--weekly|--monthly] [--date="last week"] > last-weeks-report.txt
```

(What's a good Python date/time parsing library that can parse dates like "last week"?  dateutils?)

**Comment by mgedmin** [comment]
A downside: gtimelog.main imports Gtk, which causes these messages on my Ubuntu when `$DISPLAY` is not set:
```
$ DISPLAY= python -c 'import gi; gi.require_version("Gtk", "3.0"); from gi.repository import Gtk'
gdk_mir_display_open
Failed to connect to Mir: Failed to connect to server socket: Toks failas ar aplankas neegzistuoja
Unable to init server: Nepavyko prisijungti: Prisijungimas atmestas
```
(Thank you, Canonical.)
Perhaps there could be a separate script, `gtimelog-cli`?

**Comment by standagh** [comment]
Hi All

It was a some time ago I was trying to extend gtimelog with subcategories.
I prepared patches but was not enough convincing to make Marius include
them :)

After a while I realized that all I wanted is better reporting. Editing
functions are just fine and data are in simple text file so I created
several scripts that parses / filters / summarizes tasks / time spent in
task and displays it to stdout.

Because I am using linux it's based on bash / python scripts.

Iam attaching usage instructions:
----
gtlsreport startInterval [endInterval] [project]
Interval:
    [0-9] ... how many days - 10 = for last 10 days
    +[0-9] ... how many deys in the interval. Can be used only as an
'endInterval' value
    [0-9]*. since x-th day in the month - 4. = report since the 4th of the
month till today
    [0-9]+*.[0-9]+. since x-th day in the y-th month - 4.5. = report since
the 4th of May till today
    [0-9]+*.[0-9]+.[0-9]{2} like above but includes year in 2digit form
    [0-9]+*.[0-9]+.[0-9]{4} like above but includes year in 4digit form
    w ... current week (only as 'startInterval', no 'endInterval' expected.)
    lw ... last week (only as 'startInterval', no 'endInterval' expected.)
    m ... current month (only as 'startInterval', no 'endInterval'
expected.)
    lm ... last month (only as 'startInterval', no 'endInterval' expected.)

Samples:
'gtlsreport 10' ... for last 10 days
'gtlsreport 10 +4' ... from 10th day ago and for 4 consecutive days
'gtlsreport 3. PROJ1' ... since 3th of the current month for categories
PROJ1
'gtlsreport 10.8. +5 MYOLD' ... since 10th of the August for 5 consecutive
days for category MYOLD
'gtlsreport 10.8. 22.8.15' ... since 10th of the August to 22th of the
August of 2015
'gtlsreport w' ... for current week
'gtlsreport lw HW' ... for last week for category HW
'gtlsreport m' ... for current month
'gtlsreport lm' ... for last month
----

There are total 3 reporting functions:
gtlsreport - displays summary for categories including all subcategories
gtlsreports - displays summary for categories and separately for each
subcategory
gtlsreportd - displays details for categories and separately for each
subcategory - displays all tasks separated by days

Subcategories - categories is too broad scope. I need to devide category to
subcategories. Separator is "_". So category MYOLD can have: MYOLD_PM =
project management, MYOLD_OFFICE = other administration tasks, MYOLD_PART1
= PART1 phase of the project

Sample:
'gtlsreports lm MYOLD' - displays summary for last month for category MYOLD
and subcategories like MYOLD_PM, MYOLD_OFFICE, MYOLD_PART1,...
So I see that for MYOLD I have 21hours; MYOLD_PM took 3hours of that 21hrs,
MYOLD_OFFICE took 4hrs, MYOLD_PART1 tool 12hrs...

And it also understands tags syntax and counts time in tasks with specific
tags separately.

If anybody is interested I can provide the scripts and may be it can be
helpful.

standa.




On Thu, Aug 27, 2015 at 6:04 PM, Marius Gedminas <notifications@github.com>
wrote:

> A downside: gtimelog.main imports Gtk, which causes these messages on my
> Ubuntu when $DISPLAY is not set:
>
> $ DISPLAY= python -c 'import gi; gi.require_version("Gtk", "3.0"); from gi.repository import Gtk'
> gdk_mir_display_open
> Failed to connect to Mir: Failed to connect to server socket: Toks failas ar aplankas neegzistuoja
> Unable to init server: Nepavyko prisijungti: Prisijungimas atmestas
>
> (Thank you, Canonical.)
> Perhaps there could be a separate script, gtimelog-cli?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gtimelog/gtimelog/issues/43#issuecomment-135479019>.
>



-- 
standa.

**Comment by standagh** [comment]
As I am looking at what I sent - intervals should be rather like:
    [0-9]+ ... how many days - 10 = for last 10 days
    \+[0-9]+ ... how many deys in the interval. Can be used only as an
'endInterval' value
    [0-9]+. since x-th day in the month - 4. = report since the 4th of the
month till today
    [0-9]+.[0-9]+. since x-th day in the y-th month - 4.5. = report since
the 4th of May till today
    [0-9]+.[0-9]+.[0-9]{2} like above but includes year in 2digit form
    [0-9]+.[0-9]+.[0-9]{4} like above but includes year in 4digit form
    w ... current week (only as 'startInterval', no 'endInterval' expected.)
    lw ... last week (only as 'startInterval', no 'endInterval' expected.)
    m ... current month (only as 'startInterval', no 'endInterval'
expected.)
    lm ... last month (only as 'startInterval', no 'endInterval' expected.)
...
s.


On Fri, Aug 28, 2015 at 9:57 AM, standa mitrega <standa31415@gmail.com>
wrote:

> Hi All
>
> It was a some time ago I was trying to extend gtimelog with subcategories.
> I prepared patches but was not enough convincing to make Marius include
> them :)
>
> After a while I realized that all I wanted is better reporting. Editing
> functions are just fine and data are in simple text file so I created
> several scripts that parses / filters / summarizes tasks / time spent in
> task and displays it to stdout.
>
> Because I am using linux it's based on bash / python scripts.
>
> Iam attaching usage instructions:
> ----
> gtlsreport startInterval [endInterval] [project]
> Interval:
>     [0-9] ... how many days - 10 = for last 10 days
>     +[0-9] ... how many deys in the interval. Can be used only as an
> 'endInterval' value
>     [0-9]*. since x-th day in the month - 4. = report since the 4th of the
> month till today
>     [0-9]+*.[0-9]+. since x-th day in the y-th month - 4.5. = report since
> the 4th of May till today
>     [0-9]+*.[0-9]+.[0-9]{2} like above but includes year in 2digit form
>     [0-9]+*.[0-9]+.[0-9]{4} like above but includes year in 4digit form
>     w ... current week (only as 'startInterval', no 'endInterval'
> expected.)
>     lw ... last week (only as 'startInterval', no 'endInterval' expected.)
>     m ... current month (only as 'startInterval', no 'endInterval'
> expected.)
>     lm ... last month (only as 'startInterval', no 'endInterval' expected.)
>
> Samples:
> 'gtlsreport 10' ... for last 10 days
> 'gtlsreport 10 +4' ... from 10th day ago and for 4 consecutive days
> 'gtlsreport 3. PROJ1' ... since 3th of the current month for categories
> PROJ1
> 'gtlsreport 10.8. +5 MYOLD' ... since 10th of the August for 5 consecutive
> days for category MYOLD
> 'gtlsreport 10.8. 22.8.15' ... since 10th of the August to 22th of the
> August of 2015
> 'gtlsreport w' ... for current week
> 'gtlsreport lw HW' ... for last week for category HW
> 'gtlsreport m' ... for current month
> 'gtlsreport lm' ... for last month
> ----
>
> There are total 3 reporting functions:
> gtlsreport - displays summary for categories including all subcategories
> gtlsreports - displays summary for categories and separately for each
> subcategory
> gtlsreportd - displays details for categories and separately for each
> subcategory - displays all tasks separated by days
>
> Subcategories - categories is too broad scope. I need to devide category
> to subcategories. Separator is "_". So category MYOLD can have: MYOLD_PM =
> project management, MYOLD_OFFICE = other administration tasks, MYOLD_PART1
> = PART1 phase of the project
>
> Sample:
> 'gtlsreports lm MYOLD' - displays summary for last month for category
> MYOLD and subcategories like MYOLD_PM, MYOLD_OFFICE, MYOLD_PART1,...
> So I see that for MYOLD I have 21hours; MYOLD_PM took 3hours of that
> 21hrs, MYOLD_OFFICE took 4hrs, MYOLD_PART1 tool 12hrs...
>
> And it also understands tags syntax and counts time in tasks with specific
> tags separately.
>
> If anybody is interested I can provide the scripts and may be it can be
> helpful.
>
> standa.
>
>
>
>
> On Thu, Aug 27, 2015 at 6:04 PM, Marius Gedminas <notifications@github.com
> > wrote:
>
>> A downside: gtimelog.main imports Gtk, which causes these messages on my
>> Ubuntu when $DISPLAY is not set:
>>
>> $ DISPLAY= python -c 'import gi; gi.require_version("Gtk", "3.0"); from gi.repository import Gtk'
>> gdk_mir_display_open
>> Failed to connect to Mir: Failed to connect to server socket: Toks failas ar aplankas neegzistuoja
>> Unable to init server: Nepavyko prisijungti: Prisijungimas atmestas
>>
>> (Thank you, Canonical.)
>> Perhaps there could be a separate script, gtimelog-cli?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/gtimelog/gtimelog/issues/43#issuecomment-135479019>.
>>
>
>
>
> --
> standa.
>



-- 
standa.

**Comment by kronn** [comment]
So, reading all this, I sense that my real goal is to separate the report from a mail-client. Maybe just displaying this in the GUI, in a new window, is enough. This window could then have a "mail this" button, which would do the usual things. 
However, since this changes the way is works now, I am hesitant to just rush forward. Maybe a setting "preview reports in GUI", defaulting to false, would be enough to be backwards compatible.

A separate script for cli only would be a different beast, I fear.

@standagh, is your tool using the internals of GTimeLog?

**Comment by mgedmin** [comment]
> So, reading all this, I sense that my real goal is to separate the report from a mail-client.

In the past I've recommended setting `mailer = gedit %s` in the gtimelogrc for this, as a workaround.  (Of course, that's only suitable if you never wish to send mail.)

As for myself, I launch the mail client, then cancel the mail, when I want to just take a look.

**Comment by kronn** [comment]
Indeed, setting `mailer = gvim %s` is a "solution" for me.

However, I'll think a little if I want to put my idea into a PR. I think that both a CLI and built-in GUI reporting are viable features. Separate, of course, but still.

**Comment by standagh** [comment]
Hi

In case anybody is interested I am attaching "my" solution.
@mgedmin: no it is not using any gtimelog internals.

Put it in seperate directory, extend PATH to also include the directory and
set in gtllist path to the timelog.txt - gtimelog data file. The separator
between days is empty line in source data.

Use: gtlsreport, gtlsreports or gtlsreportd

Standa.



On Mon, Aug 31, 2015 at 6:08 PM, Matthias Viehweger <
notifications@github.com> wrote:

> Closed #43 <https://github.com/gtimelog/gtimelog/issues/43>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gtimelog/gtimelog/issues/43#event-396824755>.
>



-- 
standa.
